contracts/token/ERC20/extensions/ERC20uRWA.sol

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.26;
+
+import {IERC7943} from "../../../interfaces/IERC7943.sol";
+import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import {ERC165, IERC165} from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import {ERC20Freezable} from "./ERC20Freezable.sol";
+import {ERC20Restricted} from "./ERC20Restricted.sol";
+
+/**
+ * @dev Extension of {ERC20} according to https://eips.ethereum.org/EIPS/eip-7943[EIP-7943].
+ *
+ * Combines standard ERC-20 functionality with RWA-specific features like user restrictions,
+ * asset freezing, and forced asset transfers.
+ */
+abstract contract ERC20uRWA is ERC20, ERC165, ERC20Freezable, ERC20Restricted, IERC7943 {
+    /// @inheritdoc ERC20Restricted
+    function isUserAllowed(address user) public view virtual override(IERC7943, ERC20Restricted) returns (bool) {
+        return super.isUserAllowed(user);
+    }
+
+    /// @inheritdoc ERC165
+    function supportsInterface(bytes4 interfaceId) public view virtual override(ERC165, IERC165) returns (bool) {
+        return interfaceId == type(IERC7943).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev See {IERC7943-isTransferAllowed}.
+     *
+     * CAUTION: This function is only meant for external use. Overriding it will not apply the new checks to
+     * the internal {_update} function. Consider overriding {_update} accordingly to keep both functions in sync.
+     */
+    function isTransferAllowed(address from, address to, uint256, uint256 amount) external view virtual returns (bool) {
+        return (amount <= available(from) && isUserAllowed(from) && isUserAllowed(to));
+    }
+
+    /// @inheritdoc IERC7943
+    function getFrozen(address user, uint256) public view virtual returns (uint256 amount) {
+        return frozen(user);
+    }
+
+    /// @inheritdoc IERC7943
+    function setFrozen(address user, uint256, uint256 amount) public virtual {
+        _checkFreezer(user, amount);
+        require(amount <= balanceOf(user), ERC20InsufficientBalance(user, balanceOf(user), amount));
```

Comment:
Is it actually an issue to allow setting more than the balance of a user to frozen? Couldn't the user monitor the mempool and ddos freeze attempts?

Reply:
Yeah makes sense. I'll cap it to `Math.min(amount, balanceOf(user))`. I think it's better to follow the "no op" policy here

Reply:
going further here, what if the freezer intends to freeze more assets than is currently in the wallet (maybe they have an incoming transfer or possible future tokens). Is there an issue with this usecase?